transient debugger failure

we've seen this a lot recently

```
[info] - should trigger/continue *** FAILED ***
[info]   java.lang.AssertionError: assertion failed: expected DebugBreakEvent(DebugThreadId(1),main,/tmp/1457800313085-0/testing/debug/src/main/scala/breakpoints/Breakpoints.scala,32), found DebugThreadStartEvent(DebugThreadId(1))
```

e.g. http://fommil.com/ensime/ensime-server/268/2

--- a/api/src/main/scala/org/ensime/api/common.scala
+++ b/api/src/main/scala/org/ensime/api/common.scala
@@ -3,6 +3,14 @@
 package org.ensime.api
 
 import java.io.File
+import scala.annotation.StaticAnnotation
+
+/**
+ * Indicates that something will be removed.
+ *
+ * WORKAROUND https://issues.scala-lang.org/browse/SI-7934
+ */
+class deprecating(detail: String = "") extends StaticAnnotation
 
 sealed abstract class DeclaredAs(val symbol: scala.Symbol)
 

--- a/api/src/main/scala/org/ensime/api/config.scala
+++ b/api/src/main/scala/org/ensime/api/config.scala
@@ -21,7 +21,6 @@ case class EnsimeConfig(
     subprojects: List[EnsimeModule],
     formattingPrefs: FormattingPreferences,
     sourceMode: Boolean,
-    debugArgs: List[String],
     javaLibs: List[File],
     // WORKAROUND: https://github.com/ensime/ensime-server/issues/1042
     disableSourceMonitoring: Boolean = false,
@@ -33,22 +32,18 @@ case class EnsimeConfig(
 
   /* Proposed alternatives to the legacy wire format field names */
   def root = rootDir
-  def debugVMArgs = debugArgs
   val referenceSourceJars =
     (referenceSourceRoots ++ subprojects.flatMap(_.referenceSourceRoots)).toSet
 
   // some marshalling libs (e.g. spray-json) might not like extra vals
   val modules = subprojects.map { module => (module.name, module) }.toMap
 
-  def runtimeClasspath: Set[File] =
-    compileClasspath ++ modules.values.flatMap(_.runtimeDeps) ++ targetClasspath
-
   def compileClasspath: Set[File] = modules.values.toSet.flatMap {
     m: EnsimeModule => m.compileDeps ++ m.testDeps
   } ++ (if (sourceMode) List.empty else targetClasspath)
 
   def targetClasspath: Set[File] = modules.values.toSet.flatMap {
-    m: EnsimeModule => m.targetDirs ++ m.testTargetDirs
+    m: EnsimeModule => m.targets ++ m.testTargets
   }
 
   def allJars: Set[File] = {
@@ -64,24 +59,19 @@ case class EnsimeConfig(
 
 case class EnsimeModule(
     name: String,
-    // FIXME: deprecate target/testTarget
-    target: Option[File],
     targets: List[File],
-    testTarget: Option[File],
     testTargets: List[File],
     dependsOnModules: List[String],
     compileDeps: List[File],
-    runtimeDeps: List[File],
     testDeps: List[File],
     sourceRoots: List[File],
     docJars: List[File],
     referenceSourceRoots: List[File]
 ) {
   // only check the files, not the directories, see below
-  (compileDeps ::: runtimeDeps :::
-    testDeps ::: referenceSourceRoots).foreach { f =>
-      require(f.exists, "" + f + " is required but does not exist")
-    }
+  (compileDeps ::: testDeps ::: referenceSourceRoots).foreach { f =>
+    require(f.exists, "" + f + " is required but does not exist")
+  }
 
   /*
    Proposed alternatives to the legacy wire format field names:
@@ -89,10 +79,6 @@ case class EnsimeModule(
   def compileJars = compileDeps
   def testJars = testDeps
   def referenceSourceJars = referenceSourceRoots
-
-  // prefer these to the raw target(s) until we deprecate `target`
-  val targetDirs = targets ++ target.toIterable
-  val testTargetDirs = testTargets ++ testTarget.toIterable
 
   def dependencies(implicit config: EnsimeConfig): List[EnsimeModule] =
     dependsOnModules.map(config.modules)

--- a/api/src/main/scala/org/ensime/api/incoming.scala
+++ b/api/src/main/scala/org/ensime/api/incoming.scala
@@ -287,11 +287,6 @@ case object DebugActiveVmReq extends RpcDebuggerRequest
 /**
  * Responds with `DebugVmStatus`.
  */
-case class DebugStartReq(commandLine: String) extends RpcDebuggerRequest
-
-/**
- * Responds with `DebugVmStatus`.
- */
 case class DebugAttachReq(hostname: String, port: String) extends RpcDebuggerRequest
 
 /**

--- a/core/src/it/resources/application.conf
+++ b/core/src/it/resources/application.conf
@@ -7,6 +7,9 @@ akka {
     }
   }
 
+  # needed on AppVeyor
+  logger-startup-timeout = 30s
+
   # The default 3 second wait is far too low for most of our tests.
   # When using expectNoMsg, use an explicit pragmatic timeout.
   test {

--- a/core/src/it/scala-2.11/org/ensime/intg/ImplicitsWildcardImports.scala
+++ b/core/src/it/scala-2.11/org/ensime/intg/ImplicitsWildcardImports.scala
@@ -18,7 +18,7 @@ class ImplicitsWildcardImports extends EnsimeSpec
 
   val original = EnsimeConfigFixture.ImplicitsTestProject
 
-  "ensime-server" should "allow getting symbols also after marking implicits imported with a wildcard import" ignore {
+  "ensime-server" should "allow getting symbols also after marking implicits imported with a wildcard import" in {
     withEnsimeConfig { implicit config =>
       withTestKit { implicit testkit =>
         withProject { (project, asyncHelper) =>
@@ -29,7 +29,7 @@ class ImplicitsWildcardImports extends EnsimeSpec
 
           log.info("Getting type the first time")
           project ! SymbolAtPointReq(Left(exampleFile), 116)
-          expectMsgType[Option[SymbolInfo]].get.name should be("seconds")
+          expectMsgType[SymbolInfo].name should be("seconds")
 
           log.info("Getting implicit info")
           project ! ImplicitInfoReq(Left(exampleFile), OffsetRange(0, 121))
@@ -37,7 +37,7 @@ class ImplicitsWildcardImports extends EnsimeSpec
 
           log.info("Getting type the second time")
           project ! SymbolAtPointReq(Left(exampleFile), 116)
-          expectMsgType[Option[SymbolInfo]].get.name should be("seconds")
+          expectMsgType[SymbolInfo].name should be("seconds")
         }
       }
     }

--- a/core/src/it/scala/org/ensime/core/RefactoringHandlerSpec.scala
+++ b/core/src/it/scala/org/ensime/core/RefactoringHandlerSpec.scala
@@ -213,7 +213,7 @@ class RefactoringHandlerSpec extends EnsimeSpec
     formatted should ===(expectedContents)
   }
 
-  it should "add imports on the first line when other examples come" in {
+  it should "add imports even if none exist" in {
     withAnalyzer { (dir, analyzerRef) =>
       val file = srcFile(dir, "tmp-contents", contents(
         "package org.ensime.testing",
@@ -243,6 +243,7 @@ class RefactoringHandlerSpec extends EnsimeSpec
         "package org.ensime.testing",
         "",
         "import java.lang.Integer",
+        "",
         "trait Temp {",
         "  valueOf(5)",
         "  vo(\"5\")",

--- a/core/src/it/scala/org/ensime/core/RichPresentationCompilerSpec.scala
+++ b/core/src/it/scala/org/ensime/core/RichPresentationCompilerSpec.scala
@@ -557,12 +557,12 @@ class RichPresentationCompilerSpec extends EnsimeSpec
 
     compileScala(
       List(defsFile.path),
-      config.subprojects.head.targetDirs.head,
+      config.subprojects.head.targets.head,
       cc.settings.classpath.value
     )
 
     cc.search.refreshResolver()
-    Await.result(cc.search.refresh(), Duration.Inf)
+    Await.result(cc.search.refresh(), 180.seconds)
 
     val scalaVersion = scala.util.Properties.versionNumberString
     val parts = scalaVersion.split("\\.").map { _.toInt }

--- a/core/src/it/scala/org/ensime/fixture/EnsimeConfigFixture.scala
+++ b/core/src/it/scala/org/ensime/fixture/EnsimeConfigFixture.scala
@@ -117,9 +117,7 @@ object EnsimeConfigFixture {
     // I tried using shapeless everywhere here, but it OOMd the compiler :-(
 
     def cloneModule(m: EnsimeModule): EnsimeModule = m.copy(
-      target = m.target.map(renameAndCopyTarget),
       targets = m.targets.map(renameAndCopyTarget),
-      testTarget = m.testTarget.map(renameAndCopyTarget),
       testTargets = m.testTargets.map(renameAndCopyTarget),
       sourceRoots = m.sourceRoots.map(renameAndCopy)
     )

--- a/core/src/it/scala/org/ensime/fixture/ProjectFixture.scala
+++ b/core/src/it/scala/org/ensime/fixture/ProjectFixture.scala
@@ -2,6 +2,7 @@
 // Licence: http://www.gnu.org/licenses/gpl-3.0.en.html
 package org.ensime.fixture
 
+import akka.actor.{ ActorRef, ActorSystem }
 import org.scalatest._
 import akka.testkit._
 
@@ -9,6 +10,22 @@ import org.ensime.api._
 import org.ensime.core._
 
 import scala.concurrent.duration._
+
+// WORKAROUND http://stackoverflow.com/questions/13420809
+object LoggingTestProbe {
+  def apply()(implicit system: ActorSystem): TestProbe = {
+    val probe = TestProbe()
+    probe.setAutoPilot(new TestActor.AutoPilot {
+      def run(sender: ActorRef, msg: Any) = {
+        val other = sender.path
+        val me = probe.ref.path
+        system.log.debug(s"AsyncHelper $me received $msg from $other")
+        this
+      }
+    })
+    probe
+  }
+}
 
 object ProjectFixture extends Matchers {
   private[fixture] def startup(
@@ -18,11 +35,15 @@ object ProjectFixture extends Matchers {
   ): (TestActorRef[Project], TestProbe) = {
     import testkit._
 
-    val probe = TestProbe()
+    val probe = LoggingTestProbe()
     probe.ignoreMsg {
       // these are too noisy for tests
       case e: SendBackgroundMessageEvent => true
       case e: DebugOutputEvent => true
+      case e: DebugThreadStartEvent => true
+      case e: DebugThreadDeathEvent => true
+      case e: DebugVmError => true
+      case DebugVMDisconnectEvent => true
       case ClearAllScalaNotesEvent => true
       case ClearAllJavaNotesEvent => true
     }

--- a/core/src/it/scala/org/ensime/indexer/SearchServiceSpec.scala
+++ b/core/src/it/scala/org/ensime/indexer/SearchServiceSpec.scala
@@ -36,7 +36,7 @@ class SearchServiceSpec extends EnsimeSpec
       val now = System.currentTimeMillis()
       for {
         m <- config.modules.values
-        r <- m.targetDirs ++ m.testTargetDirs
+        r <- m.targets ++ m.testTargets
         f <- r.tree
       } {
         // simulate a full recompile
@@ -52,7 +52,7 @@ class SearchServiceSpec extends EnsimeSpec
   it should "remove classfiles that have been deleted" in {
     withSearchService { (config, service) =>
       implicit val s = service
-      val classfile = config.subprojects.head.targetDirs.head / "org/example/Foo.class"
+      val classfile = config.subprojects.head.targets.head / "org/example/Foo.class"
 
       classfile shouldBe 'exists
 

--- a/core/src/it/scala/org/ensime/intg/DebugTest.scala
+++ b/core/src/it/scala/org/ensime/intg/DebugTest.scala
@@ -2,15 +2,21 @@
 // Licence: http://www.gnu.org/licenses/gpl-3.0.en.html
 package org.ensime.intg
 
+import akka.event.slf4j.SLF4JLogging
 import java.io.File
 
 import akka.testkit._
+import java.io.InputStream
+import java.util.Scanner
 import org.ensime.api._
 import org.ensime.core._
 import org.ensime.fixture._
 import org.ensime.util.EnsimeSpec
 import org.ensime.util.file._
 import org.scalatest.Matchers
+import scala.concurrent.{ Await, Future, Promise }
+import scala.util.{ Properties, Try }
+import scala.concurrent.duration._
 
 // must be refreshing as the tests don't clean up after themselves properly
 class DebugTest extends EnsimeSpec
@@ -69,81 +75,51 @@ class DebugTest extends EnsimeSpec
                 DebugStackFrame(1, List(
                   DebugStackLocal(0, "args", "Array[]", "java.lang.String[]")
                   ), 1, "breakpoints.Breakpoints$", "main",
-                  LineSourcePosition(`breakpointsFile`, 41), _),
+                  LineSourcePosition(`breakpointsFile`, 42), _),
                 DebugStackFrame(2, List(), 1, "breakpoints.Breakpoints", "main",
                   LineSourcePosition(`breakpointsFile`, _), _)
                 ), DebugThreadId(1), "main") =>
             }
 
-            //            val bp11 = session.addLineBreakpoint(BP_TYPENAME, 11)
             project ! DebugSetBreakReq(breakpointsFile, 11)
             expectMsg(TrueResponse)
-            //            val bp13 = session.addLineBreakpoint(BP_TYPENAME, 13)
+
             project ! DebugSetBreakReq(breakpointsFile, 13)
             expectMsg(TrueResponse)
-
-            //              session.waitForBreakpointsToBeEnabled(bp11, bp13)
-
-            //              session.resumetoSuspension()
-            //              session.checkStackFrame(BP_TYPENAME, "simple1()V", 11)
-
-            asyncHelper.expectMsg(DebugBreakEvent(DebugThreadId(1), "main", breakpointsFile, 32))
 
             project ! DebugContinueReq(DebugThreadId(1))
             expectMsg(TrueResponse)
 
             asyncHelper.expectMsg(DebugBreakEvent(DebugThreadId(1), "main", breakpointsFile, 11))
 
-            //              session.resumetoSuspension()
-            //              session.checkStackFrame(BP_TYPENAME, "simple1()V", 13)
-
             project ! DebugContinueReq(DebugThreadId(1))
             expectMsg(TrueResponse)
 
             asyncHelper.expectMsg(DebugBreakEvent(DebugThreadId(1), "main", breakpointsFile, 13))
 
-            //              bp11.setEnabled(false)
             project ! DebugClearBreakReq(breakpointsFile, 11)
             expectMsg(TrueResponse)
 
-            //              session.waitForBreakpointsToBeDisabled(bp11)
-            //
-            //              session.resumetoSuspension()
             project ! DebugContinueReq(DebugThreadId(1))
             expectMsg(TrueResponse)
-            //              session.checkStackFrame(BP_TYPENAME, "simple1()V", 13)
+
             asyncHelper.expectMsg(DebugBreakEvent(DebugThreadId(1), "main", breakpointsFile, 13))
-            //
-            //              bp11.setEnabled(true); bp13.setEnabled(false)
             project ! DebugSetBreakReq(breakpointsFile, 11)
             expectMsg(TrueResponse)
             project ! DebugClearBreakReq(breakpointsFile, 13)
             expectMsg(TrueResponse)
 
-            //
-            //              session.waitForBreakpointsToBeEnabled(bp11)
-            //              session.waitForBreakpointsToBeDisabled(bp13)
-            //
-            //              session.resumetoSuspension()
-            //              session.checkStackFrame(BP_TYPENAME, "simple1()V", 11)
             project ! DebugContinueReq(DebugThreadId(1))
             expectMsg(TrueResponse)
 
             asyncHelper.expectMsg(DebugBreakEvent(DebugThreadId(1), "main", breakpointsFile, 11))
-            //
-            //              session.resumetoSuspension()
-            //              session.checkStackFrame(BP_TYPENAME, "simple1()V", 11)
             project ! DebugContinueReq(DebugThreadId(1))
             expectMsg(TrueResponse)
             asyncHelper.expectMsg(DebugBreakEvent(DebugThreadId(1), "main", breakpointsFile, 11))
-            //
+
             project ! DebugContinueReq(DebugThreadId(1))
             expectMsg(TrueResponse)
 
-            //asyncHelper.expectAsync(60 seconds, DebugVMDisconnectEvent)
-            //              session.resumeToCompletion()
-            //              bp11.delete()
-            //              bp13.delete()
           }
       }
     }
@@ -160,8 +136,6 @@ class DebugTest extends EnsimeSpec
         ) { breakpointsFile =>
             import testkit._
             val breakpointsPath = breakpointsFile.getAbsolutePath
-
-            // TODO: test listing/clearing pending breakpoints (i.e. before we connect)
 
             project ! DebugListBreakpointsReq
             expectMsgType[BreakpointList] should matchPattern {
@@ -302,31 +276,43 @@ trait DebugTestUtils {
     project ! DebugSetBreakReq(resolvedFile, breakLine)
     expectMsg(TrueResponse)
 
-    project ! DebugStartReq(className)
+    val vm = VMStarter(config, className)
+    Await.result(vm._4, (5 seconds).dilated)
+
+    project ! DebugAttachReq(vm._2, vm._3.toString)
+
     expectMsg(DebugVmSuccess())
 
     asyncHelper.expectMsg(DebugVMStartEvent)
-    asyncHelper.expectMsgType[DebugThreadStartEvent]
 
-    val expect = DebugBreakEvent(DebugThreadId(1), "main", resolvedFile, breakLine)
-    asyncHelper.expectMsg(expect)
+    val gotOnStartup = asyncHelper.expectMsgType[EnsimeServerMessage]
+    // weird! we sometimes see a duplicate break event instantly, not really expected
+    val additionalOnStartup = Try(asyncHelper.expectMsgType[EnsimeServerMessage](1 second)).toOption.toSeq
+    // but it doesn't always come through
+    (gotOnStartup +: additionalOnStartup) should contain(
+      DebugBreakEvent(DebugThreadId(1), "main", resolvedFile, breakLine)
+    )
+
     project ! DebugClearBreakReq(resolvedFile, breakLine)
     expectMsg(TrueResponse)
 
     try {
       f(resolvedFile)
     } finally {
-      project ! DebugClearAllBreaksReq
-      expectMsg(TrueResponse)
-
-      // no way to await the stopped condition so we let the app run
-      // its course on the main thread
-      project ! DebugContinueReq(DebugThreadId(1))
-      expectMsg(TrueResponse)
-      project ! DebugStopReq
-      expectMsgPF() {
-        case TrueResponse =>
-        case FalseResponse => // windows does this sometimes
+      try {
+        project ! DebugClearAllBreaksReq
+        expectMsg(TrueResponse)
+        // no way to await the stopped condition so we let the app run
+        // its course on the main thread
+        project ! DebugContinueReq(DebugThreadId(1))
+        expectMsg(TrueResponse)
+        project ! DebugStopReq
+        expectMsgPF() {
+          case TrueResponse =>
+          case FalseResponse => // windows does this sometimes
+        }
+      } finally {
+        vm._1.destroy()
       }
     }
   }
@@ -352,4 +338,50 @@ trait DebugTestUtils {
         DebugThreadId(1), "main") =>
     }
   }
+}
+
+// only for test because only the build tool really knows how to launch the JVM
+object VMStarter extends SLF4JLogging {
+
+  def logLines(src: InputStream): Future[Unit] = {
+    val promise = Promise[Unit]()
+    new Thread(new Runnable() {
+      override def run(): Unit = {
+        val sc = new Scanner(src)
+        while (sc.hasNextLine()) {
+          if (!promise.isCompleted) promise.trySuccess(())
+          log.info("DEBUGGING_PROCESS:" + sc.nextLine());
+        }
+      }
+    }).start()
+    promise.future
+  }
+
+  def java: String =
+    if (Properties.isWin) Properties.javaHome + """\bin\javaw.exe"""
+    else Properties.javaHome + "/bin/java"
+
+  def apply(config: EnsimeConfig, clazz: String): (Process, String, Int, Future[Unit]) = {
+    import collection.JavaConverters._
+
+    // would be nice to have ephemeral debug ports
+    val port = 5000 + scala.util.Random.nextInt(1000)
+
+    val classpath = (config.compileClasspath ++ config.targetClasspath).mkString(File.pathSeparator)
+    val args = Seq(
+      java,
+      "-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=" + port,
+      "-classpath", classpath,
+      clazz
+    )
+
+    val process = new ProcessBuilder(
+      args.asJava
+    ).redirectErrorStream(true).start()
+
+    val logging = logLines(process.getInputStream())
+
+    (process, "127.0.0.1", port, logging)
+  }
+
 }

--- a/core/src/main/scala/org/ensime/config/EnsimeConfigProtocol.scala
+++ b/core/src/main/scala/org/ensime/config/EnsimeConfigProtocol.scala
@@ -51,18 +51,12 @@ object EnsimeConfigProtocol {
    side-effect.
    */
   private[config] def validated(m: EnsimeModule): EnsimeModule = {
-    (m.targetDirs ++ m.testTargetDirs ++ m.sourceRoots).foreach { dir =>
+    (m.targets ++ m.testTargets ++ m.sourceRoots).foreach { dir =>
       if (!dir.exists() && !dir.isJar) {
         log.warn(s"$dir does not exist, creating")
         dir.mkdirs()
       }
     }
-    Canonised(m.copy(
-      target = None,
-      targets = m.targetDirs,
-      testTarget = None,
-      testTargets = m.testTargetDirs,
-      sourceRoots = m.sourceRoots
-    ))
+    Canonised(m)
   }
 }

--- a/core/src/main/scala/org/ensime/core/debug/DebugManager.scala
+++ b/core/src/main/scala/org/ensime/core/debug/DebugManager.scala
@@ -105,11 +105,6 @@ class DebugManager(
     broadcaster ! DebugVMDisconnectEvent
   }
 
-  def vmOptions(): List[String] = List(
-    "-classpath",
-    config.runtimeClasspath.mkString("\"", File.pathSeparator, "\"")
-  ) ++ config.debugVMArgs
-
   def withVM[T](action: (VM => T)): Option[T] = {
     maybeVM.synchronized {
       try {
@@ -237,24 +232,10 @@ class DebugManager(
     DebugVmError(1, message)
   }
 
-  def handleDebugStartReq(commandLine: String): RpcResponse = {
+  def handleDebugAttachReq(hostname: String, port: Int): RpcResponse = {
     disposeCurrentVM()
     try {
-      val vm = new VM(VmStart(commandLine), vmOptions(), self, broadcaster, sourceMap)
-      maybeVM = Some(vm)
-      vm.start()
-      DebugVmSuccess()
-    } catch {
-      case e: Exception =>
-        log.error(e, "Could not start VM")
-        handleStartupFailure(e)
-    }
-  }
-
-  def handleDebugAttachReq(hostname: String, port: String): RpcResponse = {
-    disposeCurrentVM()
-    try {
-      val vm = new VM(VmAttach(hostname, port), vmOptions(), self, broadcaster, sourceMap)
+      val vm = new VM(hostname, port, self, broadcaster, sourceMap)
       maybeVM = Some(vm)
       vm.start()
       DebugVmSuccess()
@@ -266,19 +247,14 @@ class DebugManager(
   }
 
   def fromUser: Receive = {
-    case DebugStartReq(commandLine: String) =>
-      sender ! handleDebugStartReq(commandLine)
     case DebugAttachReq(hostname, port) ⇒
-      sender ! handleDebugAttachReq(hostname, port)
+      sender ! handleDebugAttachReq(hostname, port.toInt)
     case DebugActiveVmReq =>
       sender ! handleRPCWithVM() { vm =>
         TrueResponse
       }
     case DebugStopReq =>
       sender ! handleRPCWithVM() { vm =>
-        if (vm.mode.shouldExit) {
-          vm.exit(0)
-        }
         vm.dispose()
         TrueResponse
       }

--- a/core/src/main/scala/org/ensime/core/debug/VM.scala
+++ b/core/src/main/scala/org/ensime/core/debug/VM.scala
@@ -3,59 +3,43 @@
 package org.ensime.core.debug
 
 import akka.actor.ActorRef
+import akka.event.slf4j.SLF4JLogging
 import com.sun.jdi.request.{ EventRequest, EventRequestManager }
 import com.sun.jdi._
 import org.ensime.api._
-import org.slf4j.LoggerFactory
 import org.ensime.util.file._
 
 import scala.collection.mutable.ListBuffer
 import scala.collection.{ Iterable, mutable }
 
-class VM(val mode: VmMode, vmOptions: List[String], debugManager: ActorRef, broadcaster: ActorRef, sourceMap: SourceMap) {
-  val log = LoggerFactory.getLogger("DebugVM")
-
+class VM(
+    hostname: String,
+    port: Int,
+    debugManager: ActorRef,
+    broadcaster: ActorRef,
+    sourceMap: SourceMap
+) extends SLF4JLogging {
   import scala.collection.JavaConversions._
 
   private val vm: VirtualMachine = {
-    mode match {
-      case VmStart(commandLine) ⇒
-        val connector = Bootstrap.virtualMachineManager().defaultConnector()
-        val arguments = connector.defaultArguments()
+    log.info("Attach to running vm")
 
-        val opts = arguments.get("options").value
-        val allVMOpts = (List(opts) ++ vmOptions).mkString(" ")
-        arguments.get("options").setValue(allVMOpts)
-        arguments.get("main").setValue(commandLine)
-        // set the debugged process into suspend mode so we can catch it and add
-        // breakpoints (see vm start  event), otherwise we have a race condition.
-        arguments.get("suspend").setValue("true")
+    val vmm = Bootstrap.virtualMachineManager()
+    val connector = vmm.attachingConnectors().get(0)
 
-        log.info("Using Connector: " + connector.name + " : " + connector.description())
-        log.info("Connector class: " + connector.getClass.getName)
-        log.info("Debugger VM args: " + allVMOpts)
-        log.info("Debugger program args: " + commandLine)
-        connector.launch(arguments)
-      case VmAttach(hostname, port) ⇒
-        log.info("Attach to running vm")
+    val env = connector.defaultArguments()
+    env.get("port").setValue(port.toString)
+    env.get("hostname").setValue(hostname)
 
-        val vmm = Bootstrap.virtualMachineManager()
-        val connector = vmm.attachingConnectors().get(0)
-
-        val env = connector.defaultArguments()
-        env.get("port").setValue(port)
-        env.get("hostname").setValue(hostname)
-
-        log.info("Using Connector: " + connector.name + " : " + connector.description())
-        log.info("Debugger arguments: " + env)
-        log.info("Attach to VM")
-        val vm = connector.attach(env)
-        log.info("VM: " + vm.description + ", " + vm)
-        // if the remote VM has been started in suspended state, we need to nudge it
-        // if the remote VM has been started in running state, this call seems to be a no-op
-        vm.resume()
-        vm
-    }
+    log.info("Using Connector: " + connector.name + " : " + connector.description())
+    log.info("Debugger arguments: " + env)
+    log.info("Attach to VM")
+    val vm = connector.attach(env)
+    log.info("VM: " + vm.description + ", " + vm)
+    // if the remote VM has been started in suspended state, we need to nudge it
+    // if the remote VM has been started in running state, this call seems to be a no-op
+    vm.resume()
+    vm
   }
 
   //This flag is useful for debugging but not needed during general use
@@ -85,18 +69,10 @@ class VM(val mode: VmMode, vmOptions: List[String], debugManager: ActorRef, broa
 
   private val fileToUnits = mutable.HashMap[String, mutable.HashSet[ReferenceType]]()
   private val process = vm.process()
-  private val monitor = mode match {
-    case VmAttach(_, _) => Nil
-    case VmStart(_) => List(
-      new MonitorOutput(process.getErrorStream, broadcaster),
-      new MonitorOutput(process.getInputStream, broadcaster)
-    )
-  }
   private val savedObjects = new mutable.HashMap[DebugObjectId, ObjectReference]()
 
   def start(): Unit = {
     evtQ.start()
-    monitor.foreach { _.start() }
   }
 
   def exit(exitCode: Int): Unit = {
@@ -106,7 +82,6 @@ class VM(val mode: VmMode, vmOptions: List[String], debugManager: ActorRef, broa
   def dispose() = try {
     evtQ.finished = true
     vm.dispose()
-    monitor.foreach { _.finished = true }
   } catch {
     case e: VMDisconnectedException =>
   }

--- a/core/src/main/scala/org/ensime/indexer/SearchService.scala
+++ b/core/src/main/scala/org/ensime/indexer/SearchService.scala
@@ -93,11 +93,11 @@ class SearchService(
       log.info("findBases")
       config.modules.flatMap {
         case (name, m) =>
-          m.targetDirs.flatMap {
+          m.targets.flatMap {
             case d if !d.exists() => Nil
             case d if d.isJar => List(vfs.vfile(d))
             case d => scan(vfs.vfile(d))
-          } ::: m.testTargetDirs.flatMap {
+          } ::: m.testTargets.flatMap {
             case d if !d.exists() => Nil
             case d if d.isJar => List(vfs.vfile(d))
             case d => scan(vfs.vfile(d))

--- a/core/src/test/resources/application.conf
+++ b/core/src/test/resources/application.conf
@@ -7,6 +7,9 @@ akka {
     }
   }
 
+  # needed on AppVeyor
+  logger-startup-timeout = 30s
+
   # The default 3 second wait is far too low for most of our tests.
   # When using expectNoMsg, use an explicit pragmatic timeout.
   test {

--- a/core/src/test/scala/org/ensime/config/EnsimeConfigSpec.scala
+++ b/core/src/test/scala/org/ensime/config/EnsimeConfigSpec.scala
@@ -32,12 +32,11 @@ class EnsimeConfigSpec extends EnsimeSpec {
  :root-dir "$dir"
  :cache-dir "$cache"
  :reference-source-roots ()
- :debug-args ("-Dthis=that")
  :subprojects ((:name "module1"
                 :scala-version "2.10.4"
                 :depends-on-modules ()
-                :target "$abc"
-                :test-target "$abc"
+                :targets ("$abc")
+                :test-targets ("$abc")
                 :source-roots ()
                 :reference-source-roots ()
                 :compiler-args ()
@@ -50,7 +49,6 @@ class EnsimeConfigSpec extends EnsimeSpec {
       module1.name shouldBe "module1"
       module1.dependencies shouldBe empty
       config.sourceMode shouldBe false
-      config.debugVMArgs shouldBe List("-Dthis=that")
     })
   }
 
@@ -77,7 +75,7 @@ class EnsimeConfigSpec extends EnsimeSpec {
       val module1 = config.modules("module1")
       module1.name shouldBe "module1"
       module1.dependencies shouldBe empty
-      module1.targetDirs should have size 1
+      module1.targets should have size 1
     })
   }
 
@@ -102,7 +100,6 @@ class EnsimeConfigSpec extends EnsimeSpec {
                 :scala-version "2.10.4"
                 :targets ("$abc"))))""", { implicit config =>
           config.sourceMode shouldBe sourceMode
-          config.runtimeClasspath shouldBe Set(abc)
           config.compileClasspath shouldBe (
             if (sourceMode) Set.empty else Set(abc)
           )

--- a/project/EnsimeBuild.scala
+++ b/project/EnsimeBuild.scala
@@ -141,7 +141,7 @@ object EnsimeBuild extends Build {
       unmanagedJars in Compile += JavaTools,
       EnsimeKeys.unmanagedSourceArchives += file(".").getCanonicalFile / "openjdk-langtools/openjdk6-langtools-src.zip",
       libraryDependencies ++= Seq(
-        "com.h2database" % "h2" % "1.4.190",
+        "com.h2database" % "h2" % "1.4.191",
         "com.typesafe.slick" %% "slick" % "3.1.1",
         "com.zaxxer" % "HikariCP-java6" % "2.3.13",
         // Netbeans 7.4+ needs Java 7 (7.3 only needs it at runtime)
@@ -150,13 +150,13 @@ object EnsimeBuild extends Build {
         // lucene 4.8+ needs Java 7: http://www.gossamer-threads.com/lists/lucene/general/225300
         "org.apache.lucene" % "lucene-core" % luceneVersion,
         "org.apache.lucene" % "lucene-analyzers-common" % luceneVersion,
-        "org.ow2.asm" % "asm-commons" % "5.0.4",
-        "org.ow2.asm" % "asm-util" % "5.0.4",
+        "org.ow2.asm" % "asm-commons" % "5.1",
+        "org.ow2.asm" % "asm-util" % "5.1",
         "org.scala-lang" % "scala-compiler" % scalaVersion.value,
         "org.scala-lang" % "scalap" % scalaVersion.value,
         "com.typesafe.akka" %% "akka-actor" % Sensible.akkaVersion,
         "com.typesafe.akka" %% "akka-slf4j" % Sensible.akkaVersion,
-        "org.scala-refactoring" %% "org.scala-refactoring.library" % "0.9.1-SNAPSHOT",
+        "org.scala-refactoring" %% "org.scala-refactoring.library" % "0.10.0-SNAPSHOT",
         "commons-lang" % "commons-lang" % "2.6",
         "com.googlecode.java-diff-utils" % "diffutils" % "1.3.0"
       ) ++ Sensible.testLibs("it,test") ++ Sensible.shapeless(scalaVersion.value)
@@ -215,12 +215,11 @@ object EnsimeBuild extends Build {
   )
 
   lazy val testingDocs = Project("testingDocs", file("testing/docs")).settings(
+    dependencyOverrides ++= Set("com.google.guava" % "guava" % "18.0"),
     libraryDependencies ++= Seq(
-      // specifically using ForecastIOLib version 1.5.1 for javadoc 1.8 output
       "com.github.dvdme" % "ForecastIOLib" % "1.5.1" intransitive (),
-      "com.google.guava" % "guava" % Sensible.guavaVersion intransitive (),
       "commons-io" % "commons-io" % "2.4" intransitive ()
-    )
+    ) ++ Sensible.guava
   )
 
   // java project with no scala-library

--- a/project/SensibleSettings.scala
+++ b/project/SensibleSettings.scala
@@ -105,9 +105,9 @@ object Sensible {
   val akkaVersion = "2.3.14"
   val streamsVersion = "1.0"
   val scalatestVersion = "2.2.6"
-  val logbackVersion = "1.7.16"
+  val logbackVersion = "1.7.19"
   val quasiquotesVersion = "2.0.1"
-  val guavaVersion = "18.0"
+  val guavaVersion = "19.0"
 
   val macroParadise = Seq(
     compilerPlugin("org.scalamacros" % "paradise" % "2.0.1" cross CrossVersion.full)
@@ -117,7 +117,7 @@ object Sensible {
     else Nil
   } :+ "com.chuusai" %% "shapeless" % "2.3.0"
   val logback = Seq(
-    "ch.qos.logback" % "logback-classic" % "1.1.5",
+    "ch.qos.logback" % "logback-classic" % "1.1.6",
     "org.slf4j" % "slf4j-api" % logbackVersion,
     "org.slf4j" % "jul-to-slf4j" % logbackVersion,
     "org.slf4j" % "jcl-over-slf4j" % logbackVersion
@@ -131,6 +131,7 @@ object Sensible {
   def testLibs(config: String = "test") = Seq(
     "org.scalatest" %% "scalatest" % scalatestVersion % config,
     "org.scalamock" %% "scalamock-scalatest-support" % "3.2.2" % config,
+    // scalacheck 1.13.0 is java 7+
     "org.scalacheck" %% "scalacheck" % "1.12.5" % config,
     "com.typesafe.akka" %% "akka-testkit" % akkaVersion % config,
     "com.typesafe.akka" %% "akka-slf4j" % akkaVersion % config

--- a/protocol-jerky/src/test/scala/org/ensime/jerk/JerkFormatsSpec.scala
+++ b/protocol-jerky/src/test/scala/org/ensime/jerk/JerkFormatsSpec.scala
@@ -217,11 +217,6 @@ class JerkFormatsSpec extends EnsimeSpec with SprayJsonTestSupport with EnsimeTe
     )
 
     roundtrip(
-      DebugStartReq("blah blah blah"): RpcRequest,
-      """{"typehint":"DebugStartReq","commandLine":"blah blah blah"}"""
-    )
-
-    roundtrip(
       DebugAttachReq("mylovelyhorse", "13"): RpcRequest,
       """{"typehint":"DebugAttachReq","hostname":"mylovelyhorse","port":"13"}"""
     )

--- a/protocol-swanky/src/main/scala/org/ensime/server/protocol/swank/SwankFormats.scala
+++ b/protocol-swanky/src/main/scala/org/ensime/server/protocol/swank/SwankFormats.scala
@@ -643,7 +643,6 @@ object SwankProtocolRequest {
   implicit val StructureViewReqHint = TypeHint[StructureViewReq](SexpSymbol("swank:structure-view"))
   implicit val AstAtPointReqHint = TypeHint[AstAtPointReq](SexpSymbol("swank:ast-at-point"))
   implicit val DebugActiveVmReqHint = TypeHint[DebugActiveVmReq.type](SexpSymbol("swank:debug-active-vm"))
-  implicit val DebugStartReqHint = TypeHint[DebugStartReq](SexpSymbol("swank:debug-start"))
   implicit val DebugAttachReqHint = TypeHint[DebugAttachReq](SexpSymbol("swank:debug-attach"))
   implicit val DebugStopReqHint = TypeHint[DebugStopReq.type](SexpSymbol("swank:debug-stop"))
   implicit val DebugSetBreakReqHint = TypeHint[DebugSetBreakReq](SexpSymbol("swank:debug-set-break"))
@@ -792,7 +791,6 @@ object SwankProtocolRequest {
   implicit def ExpandSelectionReqFormat = SexpFormat[ExpandSelectionReq]
   implicit def StructureViewReqFormat = SexpFormat[StructureViewReq]
   implicit def AstAtPointReqFormat = SexpFormat[AstAtPointReq]
-  implicit def DebugStartReqFormat = SexpFormat[DebugStartReq]
   implicit def DebugAttachReqFormat = SexpFormat[DebugAttachReq]
   implicit def DebugSetBreakReqFormat = SexpFormat[DebugSetBreakReq]
   implicit def DebugClearBreakReqFormat = SexpFormat[DebugClearBreakReq]
@@ -845,7 +843,6 @@ object SwankProtocolRequest {
           case s if s == StructureViewReqHint.hint => value.convertTo[StructureViewReq]
           case s if s == AstAtPointReqHint.hint => value.convertTo[AstAtPointReq]
           case s if s == DebugActiveVmReqHint.hint => DebugActiveVmReq
-          case s if s == DebugStartReqHint.hint => value.convertTo[DebugStartReq]
           case s if s == DebugAttachReqHint.hint => value.convertTo[DebugAttachReq]
           case s if s == DebugStopReqHint.hint => DebugStopReq
           case s if s == DebugSetBreakReqHint.hint => value.convertTo[DebugSetBreakReq]

--- a/protocol-swanky/src/test/scala/org/ensime/server/protocol/swank/SwankFormatsSpec.scala
+++ b/protocol-swanky/src/test/scala/org/ensime/server/protocol/swank/SwankFormatsSpec.scala
@@ -233,11 +233,6 @@ class SwankFormatsSpec extends EnsimeSpec with EnsimeTestData {
     )
 
     unmarshal(
-      """(swank:debug-start "blah blah blah")""",
-      DebugStartReq("blah blah blah"): RpcRequest
-    )
-
-    unmarshal(
       """(swank:debug-attach "mylovelyhorse" "13")""",
       DebugAttachReq("mylovelyhorse", "13"): RpcRequest
     )

--- a/testing/debug/src/main/scala/breakpoints/Breakpoints.scala
+++ b/testing/debug/src/main/scala/breakpoints/Breakpoints.scala
@@ -4,7 +4,7 @@ import Helper._
 
 class Breakpoints {
 
-  def simple1() {
+  def simple1(): Unit = {
     var x = 0
 
     while (x < 4) {
@@ -16,7 +16,7 @@ class Breakpoints {
     x = 0
   }
 
-  def fors() {
+  def fors(): Unit = {
     for (i <- 1 to 3) {
       noop(None)
       noop(None)
@@ -28,7 +28,7 @@ class Breakpoints {
     }
   }
 
-  def mainTest() {
+  def mainTest(): Unit = {
     println("in mainTest")
     simple1()
     fors()
@@ -37,7 +37,8 @@ class Breakpoints {
 
 object Breakpoints {
 
-  def main(args: Array[String]) {
+  def main(args: Array[String]): Unit = {
+    println("starting Breakpoints")
     new Breakpoints().mainTest
   }
 
@@ -45,7 +46,7 @@ object Breakpoints {
 
 object Helper {
 
-  def noop(a: Any) {
+  def noop(a: Any): Unit = {
   }
 
   def ret[B](a: B): B = {

